grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java

```diff
@@ -676,78 +677,20 @@ private CastingImpl getCasting(RoleTypeImpl role, InstanceImpl rolePlayer){
         }
     }
 
-    private void putShortcutEdges(RelationImpl relation, RelationType relationType, CastingImpl newCasting){
-        Map<RoleType, Set<Instance>> roleMap = relation.allRolePlayers();
-
-        //This ensures the latest casting is taken into account when transferring relations
-        roleMap.computeIfAbsent(newCasting.getRole(), (k) -> new HashSet<>()).add(newCasting.getRolePlayer());
-
-        Set<CastingImpl> castings = relation.getMappingCasting();
+    private void putNewShortcutEdge(Instance fromInstance, Relation toRelation, RoleType roleType){
+        boolean exists  = getTinkerPopGraph().traversal().V(fromInstance.getId().getRawValue()).
+                outE(Schema.EdgeLabel.NEW_SHORTCUT.getLabel()).
+                has(Schema.EdgeProperty.RELATION_TYPE_NAME.name(), toRelation.type().getName().getValue()).
+                has(Schema.EdgeProperty.ROLE_TYPE_NAME.name(), roleType.getName().getValue()).inV().
+                hasId(toRelation.getId()).hasNext();
 
-        for (CastingImpl fromCasting : castings) {
-            RoleType fromRole = fromCasting.getRole();
-            Instance from = fromCasting.getRolePlayer();
-
-            for (CastingImpl toCasting : castings) {
-                RoleType toRole = toCasting.getRole();
-                Instance to = toCasting.getRolePlayer();
-
-                if (from != null && to != null &&  (!fromRole.equals(toRole) || !from.equals(to)) ) {
-                    putShortcutEdge(relation, relationType, fromRole, from, toRole, to);
-                }
-            }
+        if(!exists){
+            EdgeImpl edge = addEdge(fromInstance, toRelation, Schema.EdgeLabel.NEW_SHORTCUT);
```

Comment:
I feel that the edge should point _from_ the relation, _to_ the role-player, like our `CASTING` and `ROLEPLAYER` edges. Of course, it doesn't technically make a difference!